tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/NaorPinkasOt.java

```diff
@@ -0,0 +1,323 @@
+    network.send(otherId, c.toByteArray());
+    BigInteger publicKeyZero = new BigInteger(network.receive(otherId));
+    // publicKeyOne = c / keyZero
+    BigInteger publicKeyOne = publicKeyZero.modInverse(params.getP())
+        .multiply(c);
+    byte[] messageZero = new byte[hashDigest.getDigestLength()];
+    byte[] messageOne = new byte[hashDigest.getDigestLength()];
+    BigInteger encZero = encryptMessage(publicKeyZero, messageZero);
+    BigInteger encOne = encryptMessage(publicKeyOne, messageOne);
+    network.send(otherId, encZero.toByteArray());
+    network.send(otherId, encOne.toByteArray());
+    return new Pair<>(messageZero, messageOne);
+  }
+
+  /**
+   * Completes the receiver's part of the Naor-Pinkas OT.
+   *
+   * @return The message received
+   */
+  private byte[] receiveByteOt(Boolean choiceBit) {
+    BigInteger c = new BigInteger(network.receive(otherId));
+    // Pick random element privateKey
+    BigInteger privateKey = randNum.nextBigInteger(params.getP());
+    // publicKeySigma = G^privateKey mod P
+    BigInteger publicKeySigma = params.getG().modPow(privateKey, params.getP());
+    // publicKeyNotSigma = c / publicKeySigma mod P
+    BigInteger publicKeyNotSigma = publicKeySigma.modInverse(params.getP())
+        .multiply(c);
+    if (choiceBit == false) {
+      network.send(otherId, publicKeySigma.toByteArray());
+    } else {
+      network.send(otherId, publicKeyNotSigma.toByteArray());
+    }
+    BigInteger encZero = new BigInteger(network.receive(otherId));
+    BigInteger encOne = new BigInteger(network.receive(otherId));
+    byte[] message;
+    if (choiceBit == false) {
+      message = decryptMessage(encZero, privateKey);
+    } else {
+      message = decryptMessage(encOne, privateKey);
+    }
+    return message;
+  }
+
+  /**
+   * Completes the internal Naor-Pinkas encryption.
+   *
+   * @param publicKey
+   *          The public key to encrypt with
+   * @param message
+   *          The message to encrypt
+   * @return The ciphertext
+   */
+  private BigInteger encryptMessage(BigInteger publicKey,
+      byte[] message) {
+    // Pick random element r
+    BigInteger r = randNum.nextBigInteger(params.getP());
+    // Compute encryption:
+    // encryption = g^r mod P
+    BigInteger encryption = params.getG().modPow(r, params.getP());
+    // toHash = publicKey^r mod P
+    BigInteger toHash = publicKey.modPow(r, params.getP());
+    // randomMessage = H(toHash)
+    byte[] encB = hashDigest.digest(toHash.toByteArray());
+    System.arraycopy(encB, 0, message, 0, hashDigest.getDigestLength());
+    return encryption;
+  }
+
+  /**
+   * Completes the internal Naor-Pinkas decryption.
+   *
+   * @param cipher
+   *          The ciphertext to decrypt
+   * @param privateKey
+   *          The private key to use for decryption
+   * @return The plain message
+   */
+  private byte[] decryptMessage(BigInteger cipher, BigInteger privateKey) {
```

Comment:
This does not really decrypt either ... see above comment on the `encrypt` method.

Reply:
See comment for encryptMessage above.